CONTRIBUTING.md

```diff
@@ -0,0 +1,301 @@
+"expertise", OSS experience, age, or other concern. However, the development of
+standards & reference implementations demands a high-level of rigor, adversarial
+thinking, thorough testing and risk-minimization. Any bug may cost users real
+money. That being said, we deeply welcome people contributing for the first time
+to an open source project or pick up Rust while contributing. Don't be shy,
+you'll learn.
+
+
+## Communication channels
+
+Communication about Rust Bitcoin happens primarily in #rust-bitcoin IRC chat on 
+Freenode with the logs available at <http://gnusha.org/rust-bitcoin/>
+
+Discussion about code base improvements happens in GitHub issues and on pull
+requests.
+
+Major projects are tracked [here](https://github.com/orgs/rust-bitcoin/projects).
+Major milestones are tracked [here](https://github.com/rust-bitcoin/rust-bitcoin/milestones).
+
+
+## Asking questions
+
+> **Note:** Please don't file an issue to ask a question. You'll get faster 
+> results by using the resources below.
+
+We have a dedicated developer channel on IRC, #rust-bitcoin@freenode.net where 
+you may get helpful advice if you have questions.
+
+
+## Contribution workflow
+
+The codebase is maintained using the "contributor workflow" where everyone
+without exception contributes patch proposals using "pull requests". This
+facilitates social contribution, easy testing and peer review.
+
+To contribute a patch, the workflow is a as follows:
+
+1. Fork Repository
+2. Create topic branch
+3. Commit patches
+
+In general commits should be atomic and diffs should be easy to read. For this
+reason do not mix any formatting fixes or code moves with actual code changes.
+Further, each commit, individually, should compile and pass tests, in order to
+ensure git bisect and other automated tools function properly.
+
+When adding a new feature thought must be given to the long term technical debt.
+Every new features should be covered by unit tests.
+
+When refactoring, structure your PR to make it easy to review and don't hesitate
+to split it into multiple small, focused PRs.
+
+Commits should cover both the issue fixed and the solution's rationale.
+These [guidelines](https://chris.beams.io/posts/git-commit/) should be kept in
+mind.
+
+To facilitate communication with other contributors, the project is making use
+of GitHub's "assignee" field. First check that no one is assigned and then
+comment suggesting that you're working on it. If someone is already assigned,
+don't hesitate to ask if the assigned party or previous commenters are still
+working on it if it has been awhile.
+
+
+## Branches information
+
+The main library development happens in the `master` branch. This branch must
+always compile without errors (using GitHub CI). All external contributions are
+made within PRs into this branch.
+
+Each commitment within a PR to the `master` must
+* compile without errors;
+* contain all necessary tests for the introduced functional;
+* contain all docs.
```

Comment:
To be clear, each commit must compile and pass unit tests with no errors, with every feature combination (excepting `fuzztarget`, but including compiling the fuzztests) on some reasonably recent compiler. The tip of any PR branch must also compile and pass tests with no errors on 1.29.

I don't feel very strongly about insisting that commits include their own tests. I would mildly prefer that tests be in separate commits so that it's easy to reorder them during review and check that the new tests fail without the new code.